stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingStateMachine.java

```diff
@@ -443,12 +471,78 @@ private void onError(final NextProcessingStep nextStep) {
           try {
             nextStep.run();
           } catch (final Exception ex) {
-            onError(nextStep);
+            onError(ex, nextStep);
           }
         });
   }
 
+  private void startErrorLoop(final boolean isUserCommand) {
+    if (errorHandlingPhase == ErrorHandlingPhase.NO_ERROR) {
+      errorHandlingPhase =
+          isUserCommand
+              ? ErrorHandlingPhase.USER_COMMAND_PROCESSING_FAILED
+              : ErrorHandlingPhase.PROCESSING_FAILED;
+    }
+  }
+
+  private void switchErrorPhase() {
+    errorHandlingPhase =
+        switch (errorHandlingPhase) {
+          case NO_ERROR -> ErrorHandlingPhase.NO_ERROR; // First switch is explicit
+          case PROCESSING_FAILED -> ErrorHandlingPhase.PROCESSING_ERROR_FAILED;
+          case USER_COMMAND_PROCESSING_FAILED ->
+              ErrorHandlingPhase.USER_COMMAND_PROCESSING_ERROR_FAILED;
+          case USER_COMMAND_PROCESSING_ERROR_FAILED ->
+              ErrorHandlingPhase.USER_COMMAND_REJECT_FAILED;
+          case ErrorHandlingPhase.USER_COMMAND_REJECT_FAILED ->
+              ErrorHandlingPhase.USER_COMMAND_REJECT_SIMPLE_REJECT_FAILED;
+          case PROCESSING_ERROR_FAILED, USER_COMMAND_REJECT_SIMPLE_REJECT_FAILED -> {
+            LOG.error(
+                "Failed to process command '{} {}' retries. Entering endless error loop.",
+                currentRecord,
+                metadata);
+            yield ErrorHandlingPhase.ENDLESS_ERROR_LOOP;
+          }
+          case ENDLESS_ERROR_LOOP -> ErrorHandlingPhase.ENDLESS_ERROR_LOOP;
+        };
+  }
+
+  private void tryRejectingIfUserCommand(final String errorMessage) {
+    final var rejectionReason = errorMessage != null ? errorMessage : "";
+    final ProcessingResultBuilder processingResultBuilder =
+        new BufferedProcessingResultBuilder(logStreamWriter::canWriteEvents);
+    // reset value to minimize any potential error loop that can be caused by writing the full
+    // record.
+    typedCommand.getValue().reset();
```

Comment:
:question: Relating to the initial bug, this seems to be the magic :) Do you think we lose information if we immediately reset on the first phase instead of the second or third? Do you think we lose useful diagnostic information by immediately resetting it? :thinking: 

I was initially expecting we would keep it, and write an error record as replacement in the third phase, actually. But admittedly resetting is more elegant, I'm just not sure if it would lead to reduced observability in some cases.

Reply:
If all goes well, `Engine.onProcessingError` should reject with a proper error message and full details. Since this is a fall back I think it is ok to go with the safer option. 

I just quickly checked how rejection is handled in Gateway. Guess what, the value is ignored. :smile: So I think it is ok to reset the value.

Reply:
I was thinking for exporters, though I'm not aware anyone is consuming rejections. But the rejection code makes it clear that it was an error, so hopefully that's enough to explain the semantics :+1: